### PR TITLE
Add support for cross-driver sessions

### PR DESCRIPTION
**Pull request by astarche**
### What does this Pull Request accomplish?

Add support for cross-driver sessions in grpc-device to enable RFmx methods that access or initialize from ivi ViSessions. This change allows using grpc-device sessions from the corresponding drivers in those APIs.

Steps:
1. Pass in the session resource repository for all cross-driver sessions required by a service.
2. Ensure that code taking those sessions as inputs looks them up with `access_session` on the correct repository.
3. Methods that return sessions from other drivers are "cross driver init methods": they need to register that session with the correct repository.
4. Ensure that sessions created with "cross driver init methods" are removed from the repository when the initiating session is closed (the handles are closed internally by the driver).

### Why should this Pull Request be merged?

This was flagged as an important feature by the RFmx teams and the SEBU. It allows intermixing RFmx code with functionality in other drivers.

### What testing has been done?

Ran and passed all RFmx and unit tests, including new tests for cross-driver sessions.

### CODE REVIEW

File: generated/nidaqmx/nidaqmx_service.cpp [hunk 1]
```diff
@@ -30,7 +30,9 @@ namespace nidaqmx_grpc {
       NiDAQmxLibraryInterface* library,
       ResourceRepositorySharedPtr session_repository, 
       const NiDAQmxFeatureToggles& feature_toggles)
-      : library_(library), session_repository_(session_repository), feature_toggles_(feature_toggles)
```
Comment (astarche) [inline generated/nidaqmx/nidaqmx_service.cpp:33]: Splitting these to separate lines just makes it easier to conditionally codegen additional arguments

File: generated/nirfmxinstr/nirfmxinstr_service.cpp [hunk 2]
```diff
@@ -26,8 +26,12 @@ namespace nirfmxinstr_grpc {
   NiRFmxInstrService::NiRFmxInstrService(
       NiRFmxInstrLibraryInterface* library,
       ResourceRepositorySharedPtr session_repository, 
+      ViSessionResourceRepositorySharedPtr vi_session_resource_repository,
```
Comment (astarche) [inline generated/nirfmxinstr/nirfmxinstr_service.cpp:29]: RFmxInstr is a good example of the new C++ code that (a) passes through the repository pointer and (b) implements the access/init methods.

File: source/codegen/metadata/nifake_non_ivi/functions.py [hunk 3]
```diff
@@ -10,6 +10,24 @@
         ],
         'returns': 'int32'
     },
+    'GetCrossDriverSession': {
```
Comment (astarche) [inline source/codegen/metadata/nifake_non_ivi/functions.py:13]: All of the tags to make this work are a little verbose. I wanted to keep the codegen logic a little simpler, so I didn't code up mutations to infer that "cross_driver_session" outputs should be tagged as init_methods with a Session grpc_type. Feedback welcome though.

File: source/codegen/metadata/nirfmxinstr/config.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
     'namespace_component': 'nirfmxinstr',
     'close_function': 'Close',
     'custom_types': [],
-    'additional_headers': {},
+    'additional_headers': { "ivi.h": ["service.h"] },
```
Comment (astarche) [inline source/codegen/metadata/nirfmxinstr/config.py:12]: For `ViSession`

File: source/codegen/templates/service_helpers.mako [hunk 5]
```diff
@@ -39,6 +39,34 @@ ${initialize_output_params(output_parameters_to_initialize)}\
...
 </%def>
 
+<%def name="define_cross_driver_init_method_body(function_name, function_data, parameters)">\
+<%
+  initiating_driver_input_param = next(p for p in parameters if common_helpers.is_input_parameter(p) and p['grpc_type'] == 'nidevice_grpc.Session')
+  output_parameters = [p for p in parameters if common_helpers.is_output_parameter(p)]
+  session_output_param = next(p for p in output_parameters if p['grpc_type'] == 'nidevice_grpc.Session')
+  output_parameters_to_initialize = [p for p in output_parameters if p['grpc_type'] != 'nidevice_grpc.Session']
+  cross_driver_dep = service_helpers.get_cross_driver_session_dependency(session_output_param)
+  session_output_var_name = common_helpers.camel_to_snake(session_output_param['cppName'])
```
Comment (astarche) [inline source/codegen/templates/service_helpers.mako:49]: TODO: merge with #460, No raw `camel_to_snake`.

File: source/codegen/metadata/nifake_non_ivi/functions.py (hunk 3 above)
Comment (gregstoll) [reply source/codegen/metadata/nifake_non_ivi/functions.py:13]: Yeah, we're not talking about a lot of functions like this so I think I like the more verbose approach to keep the codegen simpler.

Comment (gregstoll) [reply source/codegen/metadata/nifake_non_ivi/functions.py:13]: One thing to think about is if we can add some metadata validation rules to this - for example, if a parameter has a `cross_driver_session` then the function should be an `init_method` and things like that. At least then we will know when we get the boilerplate wrong...

**Review by gregstoll — APPROVED** [review]

**Review by reckenro — APPROVED** [review]

**Pull request merged**